File: src/ai_governance/api/routers/ontology_sync.py
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status

from ai_governance.api.dependencies import get_ontology_sync_event_service
from ai_governance.api.dependencies.tenancy import get_compatible_tenant_context
from ai_governance.api.models import (
    ErrorResponse,
    OntologySyncEventListResponse,
    OntologySyncEventResponse,
    OntologySyncMetricsResponse,
)
from ai_governance.ontology.synchronization import (
    OntologySyncEvent,
    OntologySyncEventFilter,
    OntologySyncEventService,
    OntologySyncEventStatus,
    OntologySyncMetrics,
)

router = APIRouter(
    prefix="/api/v1/ontology/synchronization/events",
    tags=["Ontology Synchronization"],
)


@router.get(
    "",
    response_model=OntologySyncEventListResponse,
    status_code=status.HTTP_200_OK,
    responses={status.HTTP_500_INTERNAL_SERVER_ERROR: {"model": ErrorResponse}},
    summary="List ontology sync events",
    description="List newest-first ontology synchronization events with bounded pagination and optional filters.",
)
def list_events(
    service: Annotated[
        OntologySyncEventService,
        Depends(get_ontology_sync_event_service),
    ],
    status_filter: Annotated[
        OntologySyncEventStatus | None,
        Query(alias="status"),
    ] = None,
    entity_type: str | None = None,
    entity_id: str | None = None,
    correlation_id: str | None = None,
    event_type: str | None = None,
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
    context=Depends(get_compatible_tenant_context),
) -> OntologySyncEventListResponse:
    events = service.list_events(
        OntologySyncEventFilter(
            status=status_filter,
            entity_type=entity_type,
            entity_id=entity_id,
            correlation_id=correlation_id,
            event_type=event_type,
            organization_id=context.organization_id,
            project_id=context.project_id,
        ),
        limit=limit + 1,
        offset=offset,
    )
    return OntologySyncEventListResponse(
        events=[_to_response(event) for event in events[:limit]],
        limit=limit,
        offset=offset,
        has_more=len(events) > limit,
    )


@router.get(
    "/metrics",
    response_model=OntologySyncMetricsResponse,
    status_code=status.HTTP_200_OK,
    responses={status.HTTP_500_INTERNAL_SERVER_ERROR: {"model": ErrorResponse}},
    summary="Get ontology sync metrics",
    description="Return aggregate synchronization event metrics.",
)
def get_metrics(
    service: Annotated[
        OntologySyncEventService,
        Depends(get_ontology_sync_event_service),
    ],
    context=Depends(get_compatible_tenant_context),
) -> OntologySyncMetricsResponse:
    return _metrics_response(
        service.metrics(context.organization_id, context.project_id)
    )


@router.get(
    "/{event_id}",
    response_model=OntologySyncEventResponse,
    status_code=status.HTTP_200_OK,
    responses={
        status.HTTP_404_NOT_FOUND: {"model": ErrorResponse},
        status.HTTP_500_INTERNAL_SERVER_ERROR: {"model": ErrorResponse},
    },
    summary="Get ontology sync event",
    description="Return one synchronization event and its latest status.",
)
def get_event(
    event_id: str,
    service: Annotated[
        OntologySyncEventService,
        Depends(get_ontology_sync_event_service),
    ],
    context=Depends(get_compatible_tenant_context),
) -> OntologySyncEventResponse:
    return _to_response(_get_event(service, event_id, context))


@router.get(
    "/{event_id}/status",
    response_model=OntologySyncEventResponse,
    status_code=status.HTTP_200_OK,
    responses={
        status.HTTP_404_NOT_FOUND: {"model": ErrorResponse},
        status.HTTP_500_INTERNAL_SERVER_ERROR: {"model": ErrorResponse},
    },
    summary="Get ontology sync event status",
    description="Return one synchronization event status record.",
)
def get_event_status(
    event_id: str,
    service: Annotated[
        OntologySyncEventService,
        Depends(get_ontology_sync_event_service),
    ],
    context=Depends(get_compatible_tenant_context),
) -> OntologySyncEventResponse:
    return _to_response(_get_event(service, event_id, context))


@router.post(
    "/{event_id}/retry",
    response_model=OntologySyncEventResponse,
    status_code=status.HTTP_200_OK,
    responses={
        status.HTTP_400_BAD_REQUEST: {"model": ErrorResponse},
        status.HTTP_404_NOT_FOUND: {"model": ErrorResponse},
        status.HTTP_500_INTERNAL_SERVER_ERROR: {"model": ErrorResponse},
    },
    summary="Retry ontology sync event",
    description="Requeue a failed or dead-letter synchronization event.",
)
def retry_event(
    event_id: str,
    service: Annotated[
        OntologySyncEventService,
        Depends(get_ontology_sync_event_service),
    ],
    context=Depends(get_compatible_tenant_context),
) -> OntologySyncEventResponse:
    try:
        return _to_response(
            service.retry_event(event_id, context.organization_id, context.project_id)
        )
    except KeyError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        ) from exc


@router.post(
    "/{event_id}/cancel",
    response_model=OntologySyncEventResponse,
    status_code=status.HTTP_200_OK,
    responses={
        status.HTTP_400_BAD_REQUEST: {"model": ErrorResponse},
        status.HTTP_404_NOT_FOUND: {"model": ErrorResponse},
        status.HTTP_500_INTERNAL_SERVER_ERROR: {"model": ErrorResponse},
    },
    summary="Cancel ontology sync event",
    description="Cancel a pending synchronization event.",
)
def cancel_event(
    event_id: str,
    service: Annotated[
        OntologySyncEventService,
        Depends(get_ontology_sync_event_service),
    ],
    context=Depends(get_compatible_tenant_context),
) -> OntologySyncEventResponse:
    try:
        return _to_response(
            service.cancel_event(event_id, context.organization_id, context.project_id)
        )
    except KeyError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        ) from exc


def _get_event(
    service: OntologySyncEventService,
    event_id: str,
    context,
) -> OntologySyncEvent:
    try:
        return service.get_event(event_id, context.organization_id, context.project_id)
    except KeyError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc


def _to_response(event: OntologySyncEvent) -> OntologySyncEventResponse:
    return OntologySyncEventResponse(
        event_id=event.event_id,
        event_type=event.event_type,
        entity_type=event.entity_type,
        entity_id=event.entity_id,
        scope_identifier=event.scope_identifier,
        correlation_id=event.correlation_id,
        payload=dict(event.payload),
        status=event.status,
        retry_count=event.retry_count,
        created_at=event.created_at,
        updated_at=event.updated_at,
        completed_at=event.completed_at,
        error_message=event.error_message,
        next_retry_at=event.next_retry_at,
        last_error=event.last_error,
        failed_at=event.failed_at,
        reconciliation_report=(
            dict(event.reconciliation_report)
            if event.reconciliation_report is not None
            else None
        ),
    )


def _metrics_response(
    metrics: OntologySyncMetrics,
) -> OntologySyncMetricsResponse:
    return OntologySyncMetricsResponse(**metrics.to_dict())

File: console/lib/api/ontology-sync.ts
import {
  aiGovernanceJsonRequest,
  aiGovernanceRequest,
} from "@/lib/api/client";
import type {
  OntologySyncEvent,
  OntologySyncEventDto,
  OntologySyncEventListDto,
  OntologySyncEventStatus,
  OntologySyncMetrics,
  OntologySyncMetricsDto,
} from "@/types/ontology-sync";

const EVENTS_PATH = "/api/v1/ontology/synchronization/events";

type OntologySyncEventListFilters = {
  status?: OntologySyncEventStatus;
  entityType?: string;
  entityId?: string;
  limit?: number;
  offset?: number;
};

export async function listOntologySyncEventPage(
  filters: OntologySyncEventListFilters = {},
) {
  const dto = await aiGovernanceRequest<OntologySyncEventListDto>(EVENTS_PATH, {
    status: filters.status,
    entity_type: filters.entityType,
    entity_id: filters.entityId,
    limit: filters.limit ?? 25,
    offset: filters.offset ?? 0,
  });
  return {
    events: dto.events.map(mapEvent),
    limit: dto.limit,
    offset: dto.offset,
    hasMore: dto.has_more,
  };
}

export async function listOntologySyncEvents(
  filters: OntologySyncEventListFilters = {},
) {
  return (await listOntologySyncEventPage(filters)).events;
}

export async function getOntologySyncEvent(eventId: string) {
  const dto = await aiGovernanceRequest<OntologySyncEventDto>(
    `${EVENTS_PATH}/${encodeURIComponent(eventId)}`,
  );
  return mapEvent(dto);
}

export async function getOntologySyncMetrics() {
  const dto = await aiGovernanceRequest<OntologySyncMetricsDto>(
    `${EVENTS_PATH}/metrics`,
  );
  return mapMetrics(dto);
}

export async function retryOntologySyncEvent(eventId: string) {
  const dto = await aiGovernanceJsonRequest<OntologySyncEventDto, undefined>(
    `${EVENTS_PATH}/${encodeURIComponent(eventId)}/retry`,
    { method: "POST" },
  );
  return mapEvent(dto);
}

function mapEvent(dto: OntologySyncEventDto): OntologySyncEvent {
  return {
    eventId: dto.event_id,
    eventType: dto.event_type,
    entityType: dto.entity_type,
    entityId: dto.entity_id,
    scopeIdentifier: dto.scope_identifier,
    correlationId: dto.correlation_id,
    payload: dto.payload,
    status: dto.status,
    retryCount: dto.retry_count,
    createdAt: dto.created_at,
    updatedAt: dto.updated_at,
    completedAt: dto.completed_at,
    errorMessage: dto.error_message,
    nextRetryAt: dto.next_retry_at,
    lastError: dto.last_error,
    failedAt: dto.failed_at,
    reconciliationReport: dto.reconciliation_report,
  };
}

function mapMetrics(dto: OntologySyncMetricsDto): OntologySyncMetrics {
  return {
    totalEvents: dto.total_events,
    pendingEvents: dto.pending_events,
    processingEvents: dto.processing_events,
    completedEvents: dto.completed_events,
    failedEvents: dto.failed_events,
    deadLetterEvents: dto.dead_letter_events,
    cancelledEvents: dto.cancelled_events,
  };
}
